### PR TITLE
ipc4: invalidate cache for set large config message

### PR DESCRIPTION
**Pull request by abonislawski**
This will fix cache issue when several messages were sent one after another

### CODE REVIEW

**Comment by RanderWang** [comment]
the cache is invalided in set_large_config in each module, .e.g. copier. If you want to invalid it here, please remove all the cache invalidation in each modules.

**Comment by abonislawski** [comment]
Its not even directly in copier set_large_config but repeated in each param function (copier_set_sink_fmt, set_attenuation), is there any reason for this? For me it looks much better to make single invalidate before set_large_config() call

> static int copier_set_large_config()
> {
> 	switch (param_id) {
> 	case IPC4_COPIER_MODULE_CFG_PARAM_SET_SINK_FORMAT:
> 		return copier_set_sink_fmt(dev, data, data_offset);
> 	case IPC4_COPIER_MODULE_CFG_ATTENUATION:
> 		return set_attenuation(dev, data_offset, data);
> 	default:
> 		return -EINVAL;
> 	}
> }

**Review by kv2019i — CHANGES_REQUESTED** [review]
Looks like a a good fix, but not sure if the implementation is correct.

PS We should also move to native interface z_xtensa_cache_flush_inv(), but better to keep one API in one file, so this can be done in separate PR.

File: src/ipc/ipc4/handler.c [hunk 1]
```diff
@@ -726,6 +726,7 @@ static int ipc4_set_large_config_module_instance(struct ipc4_message_request *ip
 	const struct comp_driver *drv;
```
Comment (dbaluta) [inline src/ipc/ipc4/handler.c:726]: can you better describe the situation. what is the actual cache issue?

File: src/ipc/ipc4/handler.c [hunk 2]
```diff
@@ -726,6 +726,7 @@ static int ipc4_set_large_config_module_instance(struct ipc4_message_request *ip
...
 	int ret;
 
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)MAILBOX_HOSTBOX_BASE, config.extension.r.data_off_size);
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/ipc/ipc4/handler.c]: Well, of course it should be after 730 :) Fixed

The rest is fine, I want to invalidate MAILBOX based on config value (size)

File: src/ipc/ipc4/handler.c [hunk 3]
```diff
@@ -727,6 +727,7 @@ static int ipc4_set_large_config_module_instance(struct ipc4_message_request *ip
 	int ret;
 
 	memcpy_s(&config, sizeof(config), ipc4, sizeof(config));
+	dcache_invalidate_region((__sparse_force void __sparse_cache *)MAILBOX_HOSTBOX_BASE, config.extension.r.data_off_size);
```
Comment (lyakh) [inline src/ipc/ipc4/handler.c:730]: sorry, don't understand. `MAILBOX_HOSTBOX_BASE` is where the response is written, right? `ipc4_set_large_config_module_instance()` is just a normal IPC processing function, why does it need to invalidate cache on that area and other IPC processing functions don't need to do that?

Comment (abonislawski) [reply src/ipc/ipc4/handler.c:730]: Its not response, there is data for component set_large_config
ret = drv->ops.set_large_config(..., (char *)MAILBOX_HOSTBOX_BASE);

and this is not unique case, in the same file all cases with MAILBOX_HOSTBOX_BASE will invalidate too

**Comment by RanderWang** [comment]
> Its not even directly in copier set_large_config but repeated in each param function (copier_set_sink_fmt, set_attenuation), is there any reason for this? For me it looks much better to make single invalidate before set_large_config() call
> 
> > static int copier_set_large_config()
> > {
> > switch (param_id) {
> > case IPC4_COPIER_MODULE_CFG_PARAM_SET_SINK_FORMAT:
> > return copier_set_sink_fmt(dev, data, data_offset);
> > case IPC4_COPIER_MODULE_CFG_ATTENUATION:
> > return set_attenuation(dev, data_offset, data);
> > default:
> > return -EINVAL;
> > }
> > }

At first I thought the data size can be gotten in each module, but it can also be gotten from config.extension.r.data_off_size, so it is good to do as this PR. Please remove all the cache invalidation in set large config in each module for this PR.

**Comment by lgirdwood** [comment]
> Please remove all the cache invalidation in set large config in each module for this PR.

Ah, good point - I bet we are doing this in more than one place. @abonislawski can you check ?

**Comment by abonislawski** [comment]
Deleted invalidate entries from copier & mix_in/out

**Review by lgirdwood — APPROVED** [review]
Zephyr has no generic cache API atm, it has an xtensa specific API that needs be be made arch generic.
@kv2019i good now ?

**Review by RanderWang — APPROVED** [review]

**Review by kv2019i — APPROVED** [review]

**Pull request merged**